### ISSUE TITLE
Loading State Failed

### ISSUE THREAD

**Issue by blindsamurai777** [bookbonobo/libby-download-extension#206]
Just dropping in to say the Loading State Failed is happening again. All books. All instances. OverDrive/Libby made a back-end change to close the recent fix @ahxxm developed. Hoping someone can take a look and implement an easy fix.

**Comment by lexi1507** [comment]
Same thing happening to me, hoping for a fix soon

**Comment by DresdenErikson** [comment]
Yep, same issue for me. Fingers crossed for a fix, but I understand why it might take a while.

**Comment by LittleRedZen** [comment]
Same for me. I hope a fix is imminent, this is one of the most useful apps ever.

**Comment by MerlinGuy** [comment]
I've never worked with FF extensions before but I decided to look into this.  I stepped through the code and it is failing in background.ts - handleTitle(details: { url?: string | URL; method?: string; requestId?: any; })

While parsing the TOC the Title has a link to overdrive terms and conditions instead of the book title.  Instead of a "path" to the TOC it has the text "_d/record_access_violation"

Not good.  This is not another format change like last time.  Overdrive server knows there are web requests being made that are not passing their security checks.

Hope this helps.

**Comment by EverythingIsDucky** [comment]
FYI, I have never worked with Firefox extensions before but I thought I would look into this error.  It is failing inside background.ts - handleTitle call to parseTOC.  The Overdrive server is returning a link to their "Terms and Conditions" instead of a Title and instead of a URL path it has "-d/access violation".  Obviously the et for the book TOC is not valid and OD server code has caught this.

**Comment by ojbucao** [comment]
<img width="229" alt="Screenshot 2024-05-10 at 4 07 54 PM" src="https://github.com/bookbonobo/libby-download-extension/assets/9256675/661dc94f-9132-4810-938d-da0c08943e83">

**Comment by BLS-Pro** [comment]
Experiencing the same error as well.  Interestingly, on a very few occasions it will get past the above loading state error/failed and give that error here instead (can't screenshot my popup):

# | File | Task | Status
-- | -- | -- | --
1 |   | Loading State | Failed
2 | Part01 | Download | Running
3 | Terms & Conditions: https://company.cdn.overdrive.com/policies/terms-and-conditions | Processing | Running
4 |   | Waiting for State | Completed
5 |   | Reloading Tab | Completed

**Comment by mdf1963** [comment]
Same here.  Will help support anyone with the skill and time to create a fix.  This extension is so valuable!!

**Comment by lexi1507** [comment]
Anyone find a solution or alternate resource?

**Comment by czargamingco** [comment]
My experience is lining up with all your reports. I have been able to get the same logs to trigger at least once. But they are always failing out.

**Comment by SpeakingOfBrad** [comment]
I found how to manually download the audio of the audiobook, but it does not download the table of contents or create a .cue file so it's only sort of useful.  
I only tested this in Firefox, but I'm sure it works in Chrome too.
Open Libby, then open developer tools (F12), and go to the network tab and type .mp3 in the Filter URLs box at the top. Open your audiobook and scrub through your audiobook. As you do you'll see the Part1.mp3 Part2.mp3 etc show up. Then right-click each mp3 and open in new tab, then right-click the player bar that opens and Save audio as...

Once you have all the parts of the audiobook you can probably use tools like [AudioBookConverter](https://github.com/yermak/AudioBookConverter) and [m4b-tool](https://github.com/sandreas/m4b-tool) to find the chapters of the book and create the .cue files.

I'd also like to mention for those who don't know that [bookbonobo has a Ko-Fi](https://ko-fi.com/bookbonobo) that you can donate to show your support.

**Comment by BLS-Pro** [comment]
One thing to note in the above instructions: when you get into the developers tool and go to the network tab, make sure that "All" is selected in those network settings.  Mine had been left on a different setting at first and, until I changed it, this did not work.  :)

**Comment by czargamingco** [comment]
I have mine set to ALL and am only seeing one result from the search. Hovering over it to read the URL it only says part 1. Opening in new tab works.

**Comment by BLS-Pro** [comment]
You still have to scrub through the entire book.

**Comment by czargamingco** [comment]
Is scrubbing just jumping the chapters?  That is what is am confused what is "part 1 " vs "part 2"  part xyz

**Comment by BLS-Pro** [comment]
Scrubbing is indeed just jumping the chapters.  Doing so should create a list of gobbledeegook file names which include the terms "part01.mp3 in them.

**Comment by czargamingco** [comment]
Thanks for helping me double confirm that. Awesome work everyone teaming up to get this looked into. Of course this happened right when my stuff came in after weeks of waiting.

**Comment by SpeakingOfBrad** [comment]
When skipping through chapters make sure to pay attention to the filenames and make sure it doesn't skip any part numbers. I had a book that had multiple part files in one chapter so I had to skip through that chapter to make sure I got all of the audio files.  
(And maybe skip through the whole of the last chapter to make sure you don't miss the very end of the book.)

**Comment by czargamingco** [comment]
Just confirmed with my test with parts1-20  all worked ask told to be scrapped.  THANKS!

**Comment by Prajna-Karuna** [comment]
> Anyone find a solution or alternate resource?

The way I've been doing is the old overdrive way.
1- Locate your library on overdrive.com
2- Sign in and download the odm file. 
3- Open it in the odm in the overdrive app. (see note below)
4- Copy the mp3 files to my desired location.

Note: If you don't have the overdrive app anymore you can still download the installation file at [http://app.overdrive.com/downloads/ODMediaConsoleSetup.msi](url)
See [https://github.com/ping/odmpy/issues/69#issuecomment-2077853949](url)

**Comment by BLS-Pro** [comment]
I've never seen where one could "download the odm file".

**Comment by BLS-Pro** [comment]
"download the odm file"

Have never seen a way to do this.

**Comment by Prajna-Karuna** [comment]
Do you see the "Have OverDrive for Mac/Windows"?
If you click on the down arrow, a button will show with the option to
download the mp.

**Comment by BLS-Pro** [comment]
> Do you see the "Have OverDrive for Mac/Windows"? If you click on the down arrow, a button will show with the option to download the mp.

Yeah.  After I posted I went back and finally discovered that hidden little detail.  You helped so quickly, you apparently saw my post before I deleted it.  Thanks!  :)

**Comment by Stryfe-Delivery** [comment]
> I found how to manually download the audio of the audiobook, but it does not download the table of contents or create a .cue file so it's only sort of useful. I only tested this in Firefox, but I'm sure it works in Chrome too. Open Libby, then open developer tools (F12), and go to the network tab and type .mp3 in the Filter URLs box at the top. Open your audiobook and scrub through your audiobook. As you do you'll see the Part1.mp3 Part2.mp3 etc show up. Then right-click each mp3 and open in new tab, then right-click the player bar that opens and Save audio as...
> 
> Once you have all the parts of the audiobook you can probably use tools like [AudioBookConverter](https://github.com/yermak/AudioBookConverter) and [m4b-tool](https://github.com/sandreas/m4b-tool) to find the chapters of the book and create the .cue files.
> 
> I'd also like to mention for those who don't know that [bookbonobo has a Ko-Fi](https://ko-fi.com/bookbonobo) that you can donate to show your support.

On my end the scrub resulted in about 20 small sub-parts of each "Part 1", "Part 2", etc. the workaround here is on one of each part, right click, and click resend. This redoes the call/GET to grab the entire part. You'll see the file size go from a few kB to a few tens of MB. From there you can follow the steps to save.

**Comment by Jwink3101** [comment]
I have tried the manual way and only some files will download properly. They all play fine in Firefox but when you try to download *some* of them, they just sit and fail. I also tried on a remote computer. Is anyone else having this problem?

**Comment by audiojunky23** [comment]
> > > Anyone find a solution or alternate resource?
> > 
> > 
> > The way I've been doing is the old overdrive way. 1- Locate your library on overdrive.com 2- Sign in and download the odm file. 3- Open it in the odm in the overdrive app. (see note below) 4- Copy the mp3 files to my desired location.
> > Note: If you don't have the overdrive app anymore you can still download the installation file at http://app.overdrive.com/downloads/ODMediaConsoleSetup.msi See [ping/odmpy#69 (comment)](https://github.com/ping/odmpy/issues/69#issuecomment-2077853949)
> 
> This totally worked! Thank You! Do you bother with creating chapter .cue files?

Yeah this is where I’m at. Different method but end result is multiple mp3 files. I want to merge them into a singular file (AudioBookConverter) but I want to make a CUE file so when I open it in the Books at, it reads each chapter as a chapter and not each part or just one long “chapter”. Love to hear if people have any idea on how to do automatically

**Comment by jdishington** [comment]
The following suggestion may already have been made here -- if so, I missed it:

Audiobooks downloaded in the old OverDrive format -- consisting of multiple mp3 part-files with the chapter info for each file embedded in a custom "OverDrive" tag -- can be converted to a single-file m4b with embedded chapters (plus a cue file if you choose that option) using the free inAudible for Windows, which even provides a chapter editor.

inAudible is designed to handle Audible aax files by default. Instead, follow the path "File -> Open -> MP3 Processing" and select the folder with your OverDrive download.

The first pass converts the mp3 files to a single wav and collects the chapter info. The second pass creates the m4b and embeds the chapters into the tags and creates the optional cue file.

See https://www.epubor.com/what-is-inaudible-software.html for a safe link to download inAudible.

**Comment by lexi1507** [comment]
> The following suggestion may already have been made here -- if so, I missed it:
> 
> Audiobooks downloaded in the old OverDrive format -- consisting of multiple mp3 part-files with the chapter info for each file embedded in a custom "OverDrive" tag -- can be converted to a single-file m4b with embedded chapters (plus a cue file if you choose that option) using the free inAudible for Windows, which even provides a chapter editor.
> 
> inAudible is designed to handle Audible aax files by default. Instead, follow the path "File -> Open -> MP3 Processing" and select the folder with your OverDrive download.
> 
> The first pass converts the mp3 files to a single wav and collects the chapter info. The second pass creates the m4b and embeds the chapters into the tags and creates the optional cue file.
> 
> See https://www.epubor.com/what-is-inaudible-software.html for a safe link to download inAudible.

BLESS YOU JDISHINGTON, This worked perfectly, May you coffee or tea always be hot and may you hit every green light on all your travels :)

**Comment by jdishington** [comment]
Some tips on using inAudible to convert OverDrive audiobooks:

1. I make sure no other programs are running in Windows, as inAudible seems to work entirely in memory. Restarting Windows beforehand may be a good idea if you've been doing a lot of work.

2. I change the location of the output from its default (unfortunately the same folder as the OverDrive input files) to "Desktop", since the output filenames are based on the internal name of the audiobook and may be too long for a subfolder with an already long path. I later rename the output files to short names before moving them to their final location.

3. The longer an audiobook is, the more likely I am to split it into two or more components (as even Audible does). For example, Tolstoy's "War and Peace" is about 50 file-parts and 60 hours and 360 chapters. Splitting it into two and converting each component separately is wise. In fact necessary in some cases, because inAudible doesn't seem to embed more than 256 chapters (though the cue file is complete).

4. The cue file has an extra dummy entry at the end which I delete.

5. I've never used the chapter editor, since it would have to be done in the suspended transition from phase one to two. If I want to touch up chapter names, I edit the cue file later with a text editor (saving a copy of the original) and then use AIMP to import (re-embed) the chapters from the cue file. Alternatively, I open the m4b in the AIMP chapter editor and modify the embedded chapter tags directly, afterwards exporting them to a new cue file.

6. I've found that the chapter start times created by inAudible can be out of sync by a few seconds with the actual ones that the narrator announces (probably an artifact of combining many files into one). For example, if you select "Chapter 8" from your player's chapter list, you may find that you're already a few seconds into it or you may hear the last few seconds of the previous chapter. This was also a problem with the Firefox extension. The mismatch seems to increase gradually as the audiobook progresses, so that in a very long one it can reach 20 or 30 seconds toward the end.

**Comment by username1233414125** [comment]
Steps to extract the toc and spine:
1. There is a image/jpg item that loads in with a large amount of data included inline
2. Grab its contents, split on ",".  Get the second item in the array
3. base64 decode this, split again on "-"
4. concatenate the first and last item in resulting array
5. Take resulting item, regex replace "/\./g" -> "="
6. Regex replace again "/(.)(.)(.)(.)/g" -> "$4$3$2$1"
7. Base64 decode result.
8. This should give toc and spine.

**Comment by thomasxwright** [comment]
Yeah this is happening to me too. I hope somebody is able to find a fix for the problem!

**Comment by Mischala** [comment]
> Steps to extract the toc and spine:
>     1. There is a image/jpg item that loads in with a large amount of data included inline
>     2. Grab its contents, split on ",".  Get the second item in the array
>     3. base64 decode this, split again on "-"
>     4. concatenate the first and last item in resulting array
>     5. Take resulting item, regex replace "/./g" -> "="
>     6. Regex replace again "/(.)(.)(.)(.)/g" -> "$4$3$2$1"
>     7. Base64 decode result.
>     8. This should give toc and spine.

Thanks for this.
Works perfectly in my implementation.

**Comment by bckelly1** [comment]
> > Steps to extract the toc and spine:
> > 1. There is a image/jpg item that loads in with a large amount of data included inline
> > 2. Grab its contents, split on ",".  Get the second item in the array
> > 3. base64 decode this, split again on "-"
> > 4. concatenate the first and last item in resulting array
> > 5. Take resulting item, regex replace "/./g" -> "="
> > 6. Regex replace again "/(.)(.)(.)(.)/g" -> "$4$3$2$1"
> > 7. Base64 decode result.
> > 8. This should give toc and spine.
> 
> Thanks for this. Works perfectly in my implementation, after adjusting the replacement regex to "$4$2$3$1"

When I tried this, I get to step 3 and base64 says `X@OT&zӮ;}=Ȣpbase64: invalid input`

Not sure what I'm doing wrong. I'm using the larger of the two images in the page (favicon vs book), the content does look encoded, I'm starting with this shell command:
`cat bookname.jpg | cut -d ',' -f 2 | base64 -di`

What did you use?

**Comment by curtisgibby** [comment]
> Steps to extract the toc and spine:
> 
> 1. There is a image/jpg item that loads in with a large amount of data included inline

@username1233414125: What is the URL format of the request/response that's made? I'm not seeing it in Firefox or Chrome.

![image](https://github.com/bookbonobo/libby-download-extension/assets/1086964/301df5ba-3068-44f7-973b-ffcebc9e09d1)

**Comment by username1233414125** [comment]
@curtisgibby There is not a separate network call being made. The data is part of the HTML being returned. Look for a large inline HTML image in the header.

**Comment by username1233414125** [comment]
@bckelly1 @curtisgibby Contents start with "data:img/jpeg;base64"  If saved in n.test, can be extracted using:

`cat n.test | cut -d "," -f 2 | base64 -di | awk -F- '{ print $1$5 }' | perl -lape 's/\./=/g' | perl -lape 's/(.)(.)(.)(.)/$4$2$3$1/g' | base64 -di`

**Comment by curtisgibby** [comment]
I was able to get the TOC JSON on my Mac by removing the `-i` part of the base64 commands, i.e.

```bash
cat toc.txt | cut -d "," -f 2 | base64 -d | awk -F- '{ print $1$5 }' | perl -lape 's/\./=/g' | perl -lape 's/(.)(.)(.)(.)/$4$2$3$1/g' | base64 -d
```

(That's good!)

The JSON contains references to the actual MP3 files, but they're relative to some unknown server, so I'm unsure what the full URL to download them would be.

(That's bad!)

Is there a way to reverse-engineer what's happening in the JS in the Libby page, so that this whole process can be automated again? e.g. complete the steps 1-8 as listed here, then download the MP3s as before, package it all up as before, etc.?

**Comment by username1233414125** [comment]
If you check in the network tab, it appears that the URLS that are being fetches are like: https://dewey-16662931a0ffe78a26675bf5f70a34da.listen.libbyapp.com/%7BAA776BB5-B776-4B7B-83A4-01DD379408A7%7DFmt425-Part03.mp3?cmpt=eyJzcGluZSI6Mn0%3D--6b023bcf7fabe703c508cdb5b7577e26cbfc0d86
The chunk after the / is what you are getting from the spine decoded.  The main URL, is obtainable as seen in the network tab from this screenshot:
![image](https://github.com/user-attachments/assets/b0160031-eaf5-4439-a451-a0117d025690)

**Comment by ahxxm** [comment]
When downloading using dev console, there's no need to reverse engineer the process. Consider it this way: despite the obfuscation, the tab contains spine and actual links to the actual media files. The interaction remains click -> JavaScript -> load media files, inspect the javascript object `BIF`(book information f.......?) and go from there

**Comment by curtisgibby** [comment]
Is there any way to get this all automagically working as a Firefox extension again? Does @skittlesvampir 's fork work to correctly download all of the MP3 files?

**Comment by skittlesvampir** [comment]
No, Curtis, unfortunately my fork doesn't work anymore, as Libby is monitoring this repo and constantly breaking any fixes. I'm afraid there won't be a way to automatically download MP3s in Firefox. Merry Christmas

**Comment by ahxxm** [comment]
I received email notification today and spent sometime debugging my ~~private~~ codes to download from dev console, and found that they added some runtime params(2 parts joined by `--`?) to the media path, which I don't want to take a deeper look but could still be accessed by interacting with `BIF`, the workaround is also still sadly only possible from dev console. And due to the same big brother concern, I can only give you this hint.

**Comment by MerlinGuy** [comment]
I now just use Audacity to record the audio at night and I grab the TOC by inspecting the contents `<ul>` element and use python to parse that into a file with chapter times.  While not quick it is basically impossible to stop.

**Comment by zorin1** [comment]
> I now just use Audacity to record the audio at night and I grab the TOC by inspecting the contents `<ul>` element and use python to parse that into a file with chapter times. While not quick it is basically impossible to stop.

@MerlinGuy  What are the steps that you do to get the TOC?

**Comment by MerlinGuy** [comment]
I use the browser app so I click on the contents element to expand the chapter list.  I then use the inspect menu item to show the html.  Then select the `<UL>` element and copy the outer html.  Then I parse with python.

**Comment by kees-** [comment]
I'd been working on downloading audio with the increasing obfuscation and arrived at the same method listed above. It doesn't work to scrub and save all the media URLs created and subsequently batch download them, because they time out quickly. I'm thinking it wouldn't be a significant amount of effort to automate the process using a webcrawler with gesture support, as long as it's simple to get the media URLs as they're created. Has anyone found a way to listen for and collect the relevant network resources programmatically on the fly?

**Comment by zorin1** [comment]
> I use the browser app so I click on the contents element to expand the chapter list. I then use the inspect menu item to show the html. Then select the `<UL>` element and copy the outer html. Then I parse with python.

@MerlinGuy Thanks for the steps.  I was able to copy the UL and gave it to Deepseek and asked it to create a audacity label file.  It created the output for the file.  It's crazy how good AI have become.

**Comment by melangeaddict** [comment]
I found an alternative that worked for me this afternoon.

https://github.com/PsychedelicPalimpsest/LibbyRip

I was able to grab 4 books with minimal tinkering.  Good luck!
